tensorrt_llm/_torch/modules/fused_moe.py

```diff
@@ -700,34 +751,121 @@ def forward(
+                x, self.fc31_input_scale, 16, sfUseUE8M0, isSfSwizzledLayout)
+
+            final_hidden_states = torch.ops.trtllm.fp4_block_scale_moe_runner(
+                router_logits,
+                routing_bias,
+                hidden_states_fp4,
+                hidden_states_scale_linear_fp4.view(torch.float8_e4m3fn),
+                self.w3_w1_weight,
+                self.w3_w1_weight_scale.view(torch.float8_e4m3fn),
+                self.w2_weight,
+                self.w2_weight_scale.view(torch.float8_e4m3fn),
+                self.fc31_scale_c.data,
+                self.fc31_alpha.data,
+                self.fc2_alpha.data,
+                self.num_experts,
+                top_k,
+                n_group,
+                topk_group,
+                self.intermediate_size_per_partition,
+                self.
+                expert_start,  # local_expert_start;  use ep_rank if stride!=1
+                self.expert_size_per_partition,  # local_expert_size
+                routed_scaling_factor,
+            )
+        else:
+            raise NotImplementedError(
+                "The TRTLLM backend of FusedMoE only supports fp8_block_scaling and nvfp4 dtypes."
+            )
+
+        if self.reduce_results and self.parallel_size > 1:
+            final_hidden_states = self.all_reduce(final_hidden_states)
+
+        return final_hidden_states
+
     def load_weights(self, weights: List[Dict]):
         assert self._weights_created
         assert len(weights) == 1
         weights = weights[0]
 
-        def load_expert_w3_w1_weight(w1_weight, w3_weight,
-                                     dst_w3_w1_weight: torch.Tensor):
+        def load_expert_w3_w1_weight(w1_weight,
+                                     w3_weight,
+                                     dst_w3_w1_weight: torch.Tensor,
+                                     is_trtllm: bool = False):
             w1_weight_shard = load_weight_shard(w1_weight, self.tp_size,
                                                 self.tp_rank,
                                                 TensorParallelMode.COLUMN)
             w3_weight_shard = load_weight_shard(w3_weight, self.tp_size,
                                                 self.tp_rank,
                                                 TensorParallelMode.COLUMN)
 
-            expert_w3_weight = dst_w3_w1_weight.narrow(
-                dim=0, start=0, length=self.intermediate_size_per_partition)
-            expert_w3_weight.copy_(w3_weight_shard.view(expert_w3_weight.dtype))
+            w31_weight_shard = torch.cat([w3_weight_shard, w1_weight_shard],
+                                         dim=0)
 
-            expert_w1_weight = dst_w3_w1_weight.narrow(
-                dim=0,
-                start=self.intermediate_size_per_partition,
-                length=self.intermediate_size_per_partition)
-            expert_w1_weight.copy_(w1_weight_shard.view(expert_w1_weight.dtype))
+            if is_trtllm:
```

Comment:
Since we have `forward_cutlass` and `forward_trtllmgen`,  how about splitting `load_weights` into `load_weights_cutlass` and `load_weights_trtllmgen`?  Of course, `load_weights_cutlass` and `load_weights_trtllmgen` can share many unit functions.